keps/sig-scheduling/20200630-cost-based-scaling-down-of-pods.md

```diff
@@ -0,0 +1,382 @@
+The constraints can follow the same rules as the scheduler (through importing scheduling plugins)
+or be custom made (e.g. wrt. to application or proprietary requirements).
+The component will implement a mechanism for ranking pods.
+<!-- Either by annotating a pod, updating its status, setting a new field in pod's spec
+or creating a new CRD which will carry a cost. -->
+Each controller will have a choice to either ignore the cost or take it into account
+when scaling down.
+
+This way, the logic for selecting a victim for the scaling down operation will be
+separated from each controller. Allowing each consumer to provide its own
+logic for assigning costs. Yet, having all controllers to consume the cost uniformly.
+
+Given the default scheduler is not a source of truth about how a pod should be distributed
+after it was scheduled, scaling down strategies can exercise completely different approaches.
+
+Examples of scheduling constraints:
+- choose pods running on a node which have a `PreferNoSchedule` taint first
+- choose youngest/oldest pods first
+- choose pods minimizing topology skew among failure domains (e.g. availability zones)
+
+The goal of the proposal is not to provide specific strategies for more informed scaling down operation.
+The primary goal is to provide a mechanism and have controllers implement the mechanism.
+Allowing consumers of the new component to define their own strategies.
+
+### Examples of strategies
+
+Strategies can be divided into two categories:
+- scaling down/up a pod group does not require rank re-computation
+- scaling down/up a pod group requires rank re-computation
+
+#### Balancing duplicates among topological domains
+
+- Evict pods while minimizing skew between topological domains
+- Each pod can be given a cost based on how old/young it is in the same domain:
+  - if a pod is the first one in the domain, rank the pod with cost `1`
+  - if a pod was created second to the domain, rank the pod with cost `2`
+  - continue this way until all pods in all domains are ranked
+  - higher rank of a pod, the sooner the pod gets removed
+
+#### Pods not tolerating taints first
+
+- Evict pods that do not tolerate taints before pods that tolerate taints.
+- Each pod can be given a cost based on how many taints are not tolerated
+  - higher rank of a pod, the sooner the pod gets removed
+
+#### Minimizing pod anti-affinity
+
+- Evict pods maximizing anti-affinity first
+- Pod that improves anti-affinity on a node gets higher rank
+- Given multiple pod groups can be part of anti-affinity group, scaling down
+  a single pod in a group requires re-computation of pod ranks of all pods
+  taking part. Also, only a single pod can be scaled down at a time.
+  Otherwise, ranks no longer have to provide optimal victim selection.
+
+In the provided examples the first two strategies do not require rank re-computation.
+
+### Rank normalization and weighted sum
```

Comment:
> make the assumption that the cost of a pod is updated by a single controller

Checking something. In this proposal the cost of Pod eviction is part of its actual state (and not the `.spec` for the Pod). Is that right?

I'd expect eviction cost to be calculated by one thing using lots of plugins (much like how scoring works for scheduling), and that one thing would form part of the control plane.

Reply:
> Checking something. In this proposal the cost of Pod eviction is part of its actual state (and not the .spec for the Pod). Is that right?

Yes, it would be part of `pod.Status`.

> I'd expect eviction cost to be calculated by one thing using lots of plugins (much like how scoring works for scheduling), and that one thing would form part of the control plane.

I feel that the implementation detail of the controller updating the cost is not important here. I don't think this KEP needs to take a stand or promote implementing a controller in k/k. Such a controller warrants its own KEP if we feel we need one in-tree.